### ISSUE TITLE
Scripting the editor

### ISSUE THREAD

**Issue by Curculigo** [SpartanJ/ecode#338]
1. Replace `\n` with `)`.
2. Replace `-` with `(`.
3. Replace `\([^()]*\)` with ` `.

The input:

```
foo-1.0
bar-2.0

```

The output:

```
foo bar
```

I want to automate this boring task.

**Comment by SpartanJ** [comment]
I'll implement some Lua bindings for eepp at some point where you'll be able to implement Lua scripts using eepp resources, but that's not going to happen soon (huge task).
This feature can be implemented in several ways, it's also commonly called as "macros" in editors. Some editors like [Sublime Text](https://www.docs.sublimetext.info/en/latest/extensibility/macros.html) and [Notepad++](https://npp-user-manual.org/docs/macros/) let you record your actions and generate a macro file with the actions recorded and you can then re-run them at any time. I'm not very fond of that feature, I've never used it because it can be much more limiting than just editing a script and also much more prone to errors. I think that scriptable macros are much more powerful and I would go that way for ecode.

**Comment by Curculigo** [comment]
Scripting the whole eepp is too much. I don't want to write eepp applications in Lua (or other scripting languages). I only want to script ecode, and to be clear, only the find and replace feature. I think my regex is flawed. One could do the same thing with only one find and replace step instead of three like me. What I'm thinking now is I will write a sed script and have the build and run feature to invoke it to change the current opening file's content on the fly. Maybe easier than scripting ecode. After having to do this task more than a couple of times a day, I started to think about automating it.

**Comment by SpartanJ** [comment]
Yeah, I wasn't saying that for this particular feature I want to bind all eepp, I want to bind it for other stuff, what you need is much simpler and probably does not require even a macro.

For your particular example you can use regex captures and replace with those captured parts of the regex (you'll need to update eepp to test it, there was a small bug there with regex captures):

```
foo-1.0
bar-2.0
```

Find (for example, I don't know your real case): `(\w+)-[\d\.]+\n?`
Replace: `$1 `

$1 is the number of capture to use as replacement, this will capture the first word `(\w+)` and use it in the replacement, so result will be: `foo bar `

This is very powerful, something that is pending since forever is saving the find history per-project, that could help for some situations where some searches are commonly repeated.
Regarding macros I'll think about a good design, I don't particularly care much about it so I have no rush for this one.

**Comment by ghost** [comment]
This is to rewrite this editor in Lua to be like Lite XL? I hate Lua!

**Comment by ghost** [comment]
If you make this possible, you will find people submitting more and more small extensions in Lua and porting more and more core functionalities into Lua. You will end up with a massive Lua interpreter masquerading as an editor.

**Comment by SpartanJ** [comment]
> This is to rewrite this editor in Lua to be like Lite XL? I hate Lua!

How did you interpret that from my messages? I simply said that I want eepp to have Lua bindings at some point. That's it.

> If you make this possible, you will find people submitting more and more small extensions in Lua and porting more and more core functionalities into Lua. You will end up with a massive Lua interpreter masquerading as an editor.

I've never said that! Read carefully please, I'm talking about eepp. ecode is and will continue to be a 100% native application, performance is key.

Most probably I'll implement some command based macros (exposing some specific text document commands). But it's not decided yet how it will be approached.

One request:
* Please edit your messages, don't answer yourself, it's spammy and I end up receiving a lot of unnecessary notifications.
